### ISSUE TITLE
About self-signed certificate

### ISSUE THREAD

**Issue by onecat** [DigitalMediaServer/NSIS-INetC-plugin#3]
The old version is  ignore unknown and revoked certificates.
I using local network.
Can you change or add ignore unknown and revoked certificates?
For example: /skipsslerror

http://nsis.sourceforge.net/Inetc_plug-in
/WEAKSECURITY
 Ignore unknown and revoked certificates

**Comment by Nadahar** [comment]
`/WEAKSECURITY` is supposed to still work as before unless you're using `/NOSSL`. If that's not the case, there must be a regression. I'll have to try to test it, but I don't know any servers where I can download with an expired or invalid certificate. I guess I could install one, but I'd prefer if anybody could give me a link to test with.

**Comment by onecat** [comment]
You can view this record.
File:"NSIS-INetC-plugin/Contrib/Inetc/inetc.cpp"
* September 4, 2015 - 1.0.5.0
* By default, HTTPS connections are more secure
* Add / weak security switch to revert to the old certificate. **Security check**

Official original:
http://nsis.sourceforge.net/Inetc_plug-in
/WEAKSECURITY
Ignore unknown and **revoked certificates**

**Comment by Nadahar** [comment]
@onecat I don't think you understood me. I know what it is, and it should still work in this version. If it doesn't, that's a bug. To be able to test it, I need an URL where I can try to download from an HTTPS server with an expired or self-signed/invalid certificate.

**Comment by Nadahar** [comment]
It seems that Microsoft has abandoned everything related to the API used by this plugin. The knowledge base article dealing with it has been deleted, but I've managed to find a copy. It seems that "unknown CA" is particularly difficult to handle, and it seems to be beyond my "C/C++ skill level" to implement:




> https://support.microsoft.com/en-us/kb/182888
> 
> Retired KB Content Disclaimer
> This article was written about products for which Microsoft no longer offers support. Therefore, this article is offered "as is" and will no longer be updated.
> Summary
> If a server SSL certificate is issued by unknown or invalid certificate authority WinInet HttpSendRequest API or MFC CInternetFile::SendRequest will fail with error 12045 (ERROR_INTERNET_INVALID_CA).
> 
> When Internet Explorer tries to access the same URL, similar error is reported.
> More information
> This error occurs when the client does not know about the certificate authority that issued the server certificate. The problem may be corrected by installing the certificate authority's root certificate. A list of all installed certificates can be viewed from Internet Explorer. From the View menu, click Internet Options, click the Content tab, and click Authorities.
> 
> It is possible to bypass this error in WinInet application without installing a certificate. There are two methods of handling this error. You can use code similar to the following.
> 
> * Method 1. With a UI (a message box similar to Internet Explorer is generated):
>  ```cpp
> again: 
> if (!HttpSendRequest (hReq,...)) dwError = GetLastError(); 
> if (dwError == ERROR_INTERNET_INVALID_CA) {
>     // Make sure to check return code from InternetErrorDlg 
>     // user may click either OK or Cancel. In case of Cancel 
>     // request should not be resumbitted. 
>     InternetErrorDlg(GetDesktopWindow(), hReq, ERROR_INTERNET_INVALID_CA, FLAGS_ERROR_UI_FILTER_FOR_ERRORS | FLAGS_ERROR_UI_FLAGS_GENERATE_DATA | FLAGS_ERROR_UI_FLAGS_CHANGE_OPTIONS, NULL);
>     goto again;
> }
> ```
> * Method 2. Without a UI:
> ```cpp
> again:
> if (!HttpSendRequest(hReq,...)) dwError = GetLastError(); 
> if (dwError == ERROR_INTERNET_INVALID_CA) {
>     DWORD dwFlags;
>     DWORD dwBuffLen = sizeof(dwFlags);
>     InternetQueryOption(hReq, INTERNET_OPTION_SECURITY_FLAGS, (LPVOID)&dwFlags, &dwBuffLen);
>     dwFlags |= SECURITY_FLAG_IGNORE_UNKNOWN_CA;
>     InternetSetOption(hReq, INTERNET_OPTION_SECURITY_FLAGS, &dwFlags, sizeof (dwFlags));
>     goto again;
> }
> ```
> Similar logic can be used with MFC WinInet classes. In this case, the following MFC methods correspond to the WinInet APIs used above:
> 
>     CInternetFile::SendRequest
>     CInternetFile::QueryOption
>     CInternetFile::SetOption
>     CInternetFile::ErrorDlg 
> 
> Please note that Visual C++ 5.0 is missing documentation on CInternetFile::ErrorDlg, CInternetFile::QueryOption, and CInternetFile::SetOption. See the Inet.cpp MFC source file for information how to use this method.
> 
> NOTE 1: InternetErrorDlg may return following values: ERROR_SUCCESS ERROR_CANCELLED ERROR_INTERNET_FORCE_RETRY. The request should be resubmitted only when ERROR_INTERNET_FORCE_RETRY is returned. In Internet Explorer 4.0 and 4.01, however, the request must be resubmitted even when ERROR_SUCCESS is returned.
> 
> Microsoft has confirmed this to be a problem in InternetErrorDlg API. NOTE 2: SECURITY_FLAG_IGNORE_UNKNOWN_CA is not implemented in Internet Explorer 3.0 and 3.02.
> 
> InternetErrorDlg still works, however, with the following exception. The dialog box generated by this API does not allow ignore invalid certificate authority error; it is merely a notification to the user that page cannot be viewed.
> 
> NOTE 3: The option to ignore this error cannot be set before the error occurs. You must first attempt to send the request, receive the error, then set the option (or call InternetErrorDlg), and resubmit.
> References
> For additional information, please see the following article(s) in the Microsoft Knowledge Base:
> 168151 How to Make SSL Requests Using WinInet
> kbdsi
> Properties
> 
> Article ID: 182888 - Last Review: 06/22/2014 18:52:00 - Revision: 4.0
> Keywords:
> 
>     kberrmsg KB182888 
>

**Comment by Sami32** [comment]
https://www.kb.cert.org/vuls/id/894897/ shouldn't help neither.

**Comment by Nadahar** [comment]
@Sami32 That issue was "fixed" back in 2015 when ignoring the certificate was no longer the default.

**Comment by Sami32** [comment]
Ok, i understood that they "solved" it in ignoring this flag. But as my
english is too weak.

**Comment by Sami32** [comment]
https://stackoverflow.com/questions/23467843/sending-https-request-using-inetc

**Comment by Sami32** [comment]
https://stackoverflow.com/questions/26883305/nsis-inetc-https-gives-sendrequest-error

**Comment by Nadahar** [comment]
There is a lot of confusion regarding these flags, and I'm not sure that the SO articles are correct. From what I understand, the confusion stems from the fact that different functions accept different flags, and that the documentation is lacking. Many of the links in the documentation that is there are also dead, since Microsoft seems to have deleted vast amounts of documentation that were previously available.

This plugin uses the WinINet API, and according to the documentation, these are the valid flags for the functions of that API: https://docs.microsoft.com/en-us/windows/win32/wininet/api-flags

In the same documentation, there are something called "option flags" that can be used with some methods: https://docs.microsoft.com/en-us/windows/win32/wininet/option-flags

Some of the flags appear to be duplicates or almost duplicates. From what I could conclude, the use of the ```SECURITY_*``` flags were wrong in this context, so I replaced them with ```INTERNET_*``` flags - but I'm unsure if they exactly correspond.

Because of this, I made the following change, but unfortunately not all flags has a corresponding ```INTERNET_*``` counterpart:
https://github.com/DigitalMediaServer/NSIS-INetC-plugin/blob/ea0c45a553323f2b94ebac27f9f236ff5e25be87/Contrib/Inetc/inetc.cpp#L187-L188

After reading the documentation I found above, that has been deleted by Microsoft, it seems like it's even more complicated. It seems like it's possible to use at least some of the ```SECURITY_*``` flags, but only after first having tried and failed without. It seems to me like this is supposed to be some kind of "security" in that the user should be shown a "security dialog" and confirm that he/she wants to proceed before making the second call.

All in all, I am confused and unsure of what is actually correct here, and testing this is very tedious if you take different Windows versions and different certificate problems into account.
